src/shogun/evaluation/CrossValidation.cpp

```diff
@@ -334,33 +365,39 @@ float64_t CCrossValidation::evaluate_one_run()
 
 			/* apply machine to test features and remove subset */
 			SG_DEBUG("starting evaluation\n")
-			SG_DEBUG("%p\n", m_features)
-			CLabels* result_labels=m_machine->apply(m_features);
+			SG_DEBUG("%p\n", features)
+			CLabels* result_labels=machine->apply(features);
 			SG_DEBUG("finished evaluation\n")
-			m_features->remove_subset();
+			features->remove_subset();
 			SG_REF(result_labels);
 
 			/* evaluate */
-			results[i]=m_evaluation_criterion->evaluate(result_labels, m_labels);
+			results[i]=m_evaluation_criterion->evaluate(result_labels, labels);
 			SG_DEBUG("result on fold %d is %f\n", i, results[i])
 
 			/* evtl. update xvalidation output class */
+			#pragma omp critical
+			{
 			current=(CCrossValidationOutput*)m_xval_outputs->get_first_element();
 			while (current)
 			{
 				current->update_test_indices(subset_indices, "\t");
 				current->update_test_result(result_labels, "\t");
-				current->update_test_true_result(m_labels, "\t");
+				current->update_test_true_result(labels, "\t");
 				current->post_update_results();
 				current->update_evaluation_result(results[i], "\t");
 				SG_UNREF(current);
 				current=(CCrossValidationOutput*)
 						m_xval_outputs->get_next_element();
 			}
+			}
 
 			/* clean up, remove subsets */
+			labels->remove_subset();		
+			SG_UNREF(machine);
```

Comment:
Then you can avoid SG_REFing the m_machine (which doesnt really make sense)
And then instead, you can ref the output of the clones